File: home_assignments/csharp/TicketSeller/server/implementations/FestivalService.cs
using System;
using System.Collections.Generic;
using log4net;
using model.entities;
using model.validators;
using model.validators.interfaces;
using persistence.interfaces;
using services;
using services.interfaces;

namespace server.implementations;

public class FestivalService : IFestivalService
{
    private static readonly ILog Logger = LogManager.GetLogger("FestivalService");

    private readonly IFestivalValidator _festivalValidator;
    private readonly IFestivalRepository _festivalRepository;
    private readonly ITicketValidator _ticketValidator;
    private readonly ITicketRepository _ticketRepository;

    public FestivalService(IFestivalRepository festivalRepository, IFestivalValidator festivalValidator,
        ITicketRepository ticketRepository, ITicketValidator ticketValidator)
    {
        this._festivalValidator = festivalValidator;
        this._festivalRepository = festivalRepository;
        this._ticketValidator = ticketValidator;
        this._ticketRepository = ticketRepository;
        Logger.InfoFormat("Instantiated FestivalService with Festival repo:{0}, Festival validator:{1}," +
                          " Ticket repo:{2}, Ticket validator:{3}",
            festivalRepository, festivalValidator, ticketRepository, ticketValidator);
    }

    public ICollection<Festival> GetAllFestivals()
    {
        Logger.Info("Entered get all");
        ICollection<Festival> toReturn = _festivalRepository.GetAll();
        Logger.InfoFormat("Exit with:{0}", toReturn);
        return toReturn;
    }

    public ICollection<Festival> GetAllFestivalsOnDate(DateTime date)
    {
        Logger.InfoFormat("getting all festivals on date:{0}", date);
        try
        {
            _festivalValidator.ValidateDate(date);
        }
        catch (ValidationException e)
        {
            Logger.Error(e);
            throw new ServiceException(e.Message);
        }

        ICollection<Festival> toReturn = _festivalRepository.GetAllOnDate(date);
        Logger.InfoFormat("exiting with:{0}", toReturn);
        return toReturn;
    }

    public Ticket SellTicket(int festivalId, string buyerName, int spots)
    {
        Logger.InfoFormat("selling ticket to festival:{0} for buyer:{1} with spots:{2}",
            festivalId, buyerName, spots);
        try
        {
            _festivalValidator.ValidateId(festivalId);
        }
        catch (ValidationException e)
        {
            Logger.Error(e);
            throw new ServiceException(e.Message);
        }

        Festival festival = _festivalRepository.FindById(festivalId);
        if (festival == null)
        {
            String error = "No festival with given id found!;\n";
            Logger.Error(error);
            throw new ServiceException(error);
        }

        Ticket ticket = new Ticket(0, buyerName, festival, spots);
        try
        {
            _ticketValidator.Validate(ticket);
        }
        catch (ValidationException e)
        {
            Logger.Error(e);
            throw new ServiceException(e.Message);
        }

        _ticketRepository.Add(ticket);
        festival.SoldSpots += +spots;
        _festivalRepository.Update(festival, festival.Id);
        Logger.InfoFormat("Exiting buy method with:{0}",ticket);
        return ticket;
    }
}

File: home_assignments/java/TicketSeller/TicketSellerServer/src/main/java/ticketseller/server/FestivalService.java
package ticketseller.server;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ticketseller.model.entities.Festival;
import ticketseller.model.entities.Ticket;
import ticketseller.model.validators.interfaces.IFestivalValidator;
import ticketseller.model.validators.interfaces.ITicketValidator;
import ticketseller.persistence.interfaces.IFestivalRepository;
import ticketseller.persistence.interfaces.ITicketRepository;
import ticketseller.services.ServiceException;
import ticketseller.model.validators.ValidationException;
import ticketseller.services.interfaces.IFestivalService;

import java.time.LocalDate;
import java.util.Collection;

public class FestivalService implements IFestivalService {

    private static final Logger logger = LogManager.getLogger();

    private final IFestivalValidator festivalValidator;
    private final IFestivalRepository festivalRepository;
    private final ITicketValidator ticketValidator;
    private final ITicketRepository ticketRepository;

    public FestivalService(IFestivalRepository festivalRepository, IFestivalValidator festivalValidator,
                           ITicketRepository ticketRepository, ITicketValidator ticketValidator) {
        this.festivalValidator = festivalValidator;
        this.festivalRepository = festivalRepository;
        this.ticketValidator = ticketValidator;
        this.ticketRepository = ticketRepository;
        logger.info("Instantiated FestivalService with Festival repo:{}, Festival validator:{}," +
                        " Ticket repo:{}, Ticket validator:{}",
                festivalRepository, festivalValidator, ticketRepository, ticketValidator);
    }

    @Override
    public Collection<Festival> getAllFestivals() {
        logger.traceEntry();
        Collection<Festival> toReturn = festivalRepository.getAll();
        logger.traceExit(toReturn);
        return toReturn;
    }

    @Override
    public Collection<Festival> getAllFestivalsOnDate(LocalDate date) throws ServiceException {
        logger.traceEntry("getting all festivals on date:{}", date);
        try {
            festivalValidator.validateDate(date);
        } catch (ValidationException e) {
            logger.error(e);
            throw new ServiceException(e.getMessage());
        }
        Collection<Festival> toReturn = festivalRepository.getAllOnDate(date);
        logger.traceExit(toReturn);
        return toReturn;
    }

    @Override
    public Ticket sellTicket(Integer festivalID, String buyerName, Integer spots)
            throws ServiceException{
        logger.traceEntry("selling ticket to festival:{} for buyer:{} with spots:{}",
                festivalID, buyerName, spots);
        try {
            festivalValidator.validateID(festivalID);
        } catch (ValidationException e) {
            logger.error(e);
            throw new ServiceException(e.getMessage());
        }
        Festival festival = festivalRepository.findByID(festivalID);
        if (festival == null) {
            String error = "No festival with given id found!;\n";
            logger.error(error);
            throw new ServiceException(error);
        }
        Ticket ticket = new Ticket(null,buyerName, festival, spots);
        try {
            ticketValidator.validate(ticket);
        } catch (ValidationException e) {
            logger.error(e);
            throw new ServiceException(e.getMessage());
        }
        ticketRepository.add(ticket);
        festival.setSoldSpots(festival.getSoldSpots() + spots);
        festivalRepository.update(festival, festival.getId());
        logger.traceExit(festival);
        return ticket;
    }
}
